pyro/poutine/plate_messenger.py

```diff
@@ -19,3 +22,53 @@ def __enter__(self):
         if self._vectorized and self._indices is not None:
             return self.indices
         return None
+
+
+@contextmanager
+def unplate(name=None, dim=None):
+    """
+    EXPERIMENTAL Context manager to temporarily exit a single enclosing plate.
+
+    This is useful for sampling auxiliary variables or lazily sampling global
+    variables that are needed in a plated context. For example the following
+    models are equivalent:
+
+    Example::
+
+        def model_1(data):
+            loc = pyro.sample("loc", dist.Normal(0, 1))
+            with pyro.plate("data", len(data)):
+                with unplate("data"):
+                    scale = pyro.sample("scale", dist.LogNormal(0, 1))
+                pyro.sample("x", dist.Normal(loc, scale))
+
+        def model_2(data):
+            loc = pyro.sample("loc", dist.Normal(0, 1))
+            scale = pyro.sample("scale", dist.LogNormal(0, 1))
+            with pyro.plate("data", len(data)):
+                pyro.sample("x", dist.Normal(loc, scale))
+
+    :param str name: Optional name of plate to match.
+    :param int dim: Optional dim of plate to match. Must be negative.
+    :raises: ValueError if no enclosing plate was found.
+    """
+    if (name is not None) == (dim is not None):
+        raise ValueError("Exactly one of name,dim must be specified")
+    if name is not None:
+        assert isinstance(name, str)
+    if dim is not None:
+        assert isinstance(dim, int)
+        assert dim < 0
+
+    def predicate(messenger):
+        if not isinstance(messenger, PlateMessenger):
+            return False
+        if name is not None:
+            return messenger.name == name
+        if dim is not None:
+            return messenger.dim == dim
+
+    with mute_messengers(predicate) as matches:
+        if len(matches) != 1:
```

Comment:
When would `len(matches) > 1`?

Reply:
I guess `len(matches) in {0,1}` as written. It would be easy to generalize `name` to a regular expression, in which case this check would be more fully exercised.